File: src/algorithms/pathfinding/heuristic-search/d-star-lite/sources/DStarLite_test.java
// javac DStarLite.java DStarLite_test.java && java -ea DStarLite_test
public class DStarLite_test {

    static int[][] makeEmptyGrid(int rows, int cols) {
        return new int[rows][cols];
    }

    static void setWall(int[][] grid, int row, int col) {
        grid[row][col] = 1;
    }

    public static void main(String[] args) {
        // Test: finds path on empty grid
        {
            int[][] grid = makeEmptyGrid(5, 5);
            int[][] path = DStarLite.dStarLite(grid, new int[]{0, 0}, new int[]{4, 4});
            assert path.length > 0 : "Expected non-empty path";
            assert path[0][0] == 0 && path[0][1] == 0 : "Path should start at [0,0]";
            assert path[path.length-1][0] == 4 && path[path.length-1][1] == 4 : "Path should end at [4,4]";
        }

        // Test: returns empty path when no route
        {
            int[][] grid = makeEmptyGrid(5, 5);
            setWall(grid, 0, 1);
            setWall(grid, 1, 0);
            setWall(grid, 1, 1);
            int[][] path = DStarLite.dStarLite(grid, new int[]{0, 0}, new int[]{4, 4});
            assert path.length == 0 : "Expected empty path";
        }

        // Test: handles adjacent start and end
        {
            int[][] grid = makeEmptyGrid(3, 3);
            int[][] path = DStarLite.dStarLite(grid, new int[]{0, 0}, new int[]{0, 1});
            assert path.length > 0 : "Expected non-empty path";
            assert path[path.length-1][0] == 0 && path[path.length-1][1] == 1 : "Path should end at [0,1]";
        }

        // Test: handles start equal to end
        {
            int[][] grid = makeEmptyGrid(3, 3);
            int[][] path = DStarLite.dStarLite(grid, new int[]{1, 1}, new int[]{1, 1});
            assert path.length == 1 : "Expected path length 1";
        }

        System.out.println("All tests passed!");
    }
}

File: src/algorithms/pathfinding/heuristic-search/d-star-lite/__tests__/d-star-lite_test.go
package dstarlite

import "testing"

func makeEmptyGrid(rows, cols int) [][]GridCell {
	grid := make([][]GridCell, rows)
	for row := range grid {
		grid[row] = make([]GridCell, cols)
		for col := range grid[row] {
			grid[row][col] = GridCell{Row: row, Col: col, CellType: CellEmpty, State: "default"}
		}
	}
	return grid
}

func setWallCell(grid [][]GridCell, row, col int) {
	grid[row][col].CellType = CellWall
}

func TestFindsPathOnEmptyGrid(t *testing.T) {
	grid := makeEmptyGrid(5, 5)
	result := DStarLite(grid, 0, 0, 4, 4)
	if len(result.Path) == 0 {
		t.Error("expected non-empty path")
	}
	if result.Path[0][0] != 0 || result.Path[0][1] != 0 {
		t.Errorf("expected path start [0,0]")
	}
	last := result.Path[len(result.Path)-1]
	if last[0] != 4 || last[1] != 4 {
		t.Errorf("expected path end [4,4]")
	}
}

func TestReturnsEmptyPathWhenNoRoute(t *testing.T) {
	grid := makeEmptyGrid(5, 5)
	setWallCell(grid, 0, 1)
	setWallCell(grid, 1, 0)
	setWallCell(grid, 1, 1)
	result := DStarLite(grid, 0, 0, 4, 4)
	if len(result.Path) != 0 {
		t.Errorf("expected empty path, got %d steps", len(result.Path))
	}
}

func TestHandlesStartEqualToEnd(t *testing.T) {
	grid := makeEmptyGrid(3, 3)
	result := DStarLite(grid, 1, 1, 1, 1)
	if len(result.Path) != 1 {
		t.Errorf("expected path length 1, got %d", len(result.Path))
	}
}

func TestPerformsAtLeastOneSearch(t *testing.T) {
	grid := makeEmptyGrid(5, 5)
	result := DStarLite(grid, 0, 0, 4, 4)
	if result.ReplanCount == 0 {
		t.Error("expected replanCount > 0")
	}
}

func TestTracksVisitedCells(t *testing.T) {
	grid := makeEmptyGrid(5, 5)
	result := DStarLite(grid, 0, 0, 4, 4)
	if len(result.Visited) == 0 {
		t.Error("expected non-empty visited list")
	}
}
